dwave/plugins/torch/autoencoder/autoencoder.py

```diff
@@ -0,0 +1,133 @@
+# https://cloud.dwavesys.com/leap/legal/cloud_subscription_agreement/
+#
+
+from collections.abc import Callable
+from typing import Optional
+
+import torch
+
+__all__ = ["DiscreteAutoEncoder"]
+
+
+class DiscreteAutoEncoder(torch.nn.Module):
+    """DiscreteAutoEncoder architecture amenable for training discrete models as priors.
+    See https://iopscience.iop.org/article/10.1088/2632-2153/aba220
+
+    Such discrete models include spin-variable models amenable for the QPU. This
+    architecture is a modification of the standard autoencoder architecture, where
+    the encoder outputs a latent representation of the data, and the decoder
+    reconstructs the data from the latent representation. In our case, there is an
+    additional step where the latent representation is mapped to a discrete
+    representation, which is then passed to the decoder.
+
+    Args:
+        encoder (torch.nn.Module): The encoder must output latents that are later on
+            passed to ``latent_to_discrete``. An encoder has signature (x) -> l. x has
+            shape (batch_size, f1, f2, ...) and l has shape (batch_size, l1, l2, ...).
+        decoder (torch.nn.Module): Decodes discrete tensors into data tensors. A decoder
+            has signature (d) -> x'. d has shape (batch_size, n, d1, d2, ...) and x' has
+            shape (batch_size, f'1, f'2, ...); if x' is the reconstructed data then
+            fi=f'i, but x' might be another representation of the data (e.g. in a
+            text-to-image model, x is a sequence of tokens, and x' is an image). Note
+            that the decoder input is of shape (batch_size, n, d1, d2, ...), where n is
+            a number of discrete representations to be created from a single latent
+            representation of a single initial data point.
+        latent_to_discrete (Callable[[torch.Tensor, int], torch.Tensor] | None): A
+            stochastic and differentiable function that maps the output of the encoder
+            to a discrete representation. Importantly, since the function is stochastic,
+            it allows for the creation of multiple discrete representations from the
+            latent representation of a single data point. Thus, the signature of this
+            function is (l, n) -> d, where l is the output of the encoder and has shape
+            (batch_size, l1, l2, ...), n is the number of discrete representations per
+            data point, and d has shape (batch_size, n, d1, d2, ...), which will be the
+            input to the decoder. If None, the gumbel softmax function is used for
+            stochasticity. Defaults to None.
+    """
+
+    def __init__(
+        self,
+        encoder: torch.nn.Module,
+        decoder: torch.nn.Module,
+        latent_to_discrete: Optional[Callable[[torch.Tensor, int], torch.Tensor]] = None,
+    ):
+        super().__init__()
+        self.encoder = encoder
+        self.decoder = decoder
```

Comment:
For the sake of consistency, and some added security, let's change this to `self._encoder`, self._decoder`, and further down, `self._latent_to_discrete`, and also add properties for all.